### PR TITLE
Added tutorial for shell sort

### PR DESCRIPTION
**Pull request by Saishreekouda**
## Change Summary

- closes: #97

## Checklist

**If you haven't fulfilled the below requirements or even delete the entire checklist, your PR won't be reviewed and will be closed without notice. Regular contributors (with 10+ PRs) can skip this part.**

### General 

- [x] This Pull Request is all my own work. (You'll be blacklisted if you are caught for plagiarism.)
- [x] I've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md)
- [x] I've applied LaTex for all variables, formulas and time / space complexity instead of using backticks
- [x] I've started the app locally and verified all the content and all links (if applicable) are accessible correctly
- [x] I've included Complexity Analysis (Time Complexity & Space Complexity).
- [x] I've written my explanation well and it is easy to understand for beginners

### Tutorial

- [x] I've read and followed the [Tutorial Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#tutorial-template)
- [x] I've explained my topic well with 2 - 3 LC problems and no external problems are used. 
- [x] I've provided the full working solutions to the problems used in this tutorial.
- [x] I've provided suggested problems at the end with the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [x] I've given credits / references if I use external resources. (For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).)

### Solutions

- [x] I've read and followed the [Solution Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#solution-template)
- [x] I've formatted my code well with [K&R Coding style](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)
- [x] I've confirmed that comments are put above each line rather than writing on the same line.
- [x] I've included a meaningful approach name for my solution. e.g. `## Approach 1: Two Pointers`.

### CODE REVIEW

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. not detailed enough. expecting diagrams to explain as well.
2. please write it on your own. your overview is copied from https://www.tutorialspoint.com/data_structures_algorithms/shell_sort_algorithm.htm

**Comment by Saishreekouda** [comment]
@wingkwong I made a few changes.

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 1]
```diff
@@ -354,4 +354,60 @@ class Solution:
...
 ```
 </TabItem>
-</Tabs>
\ No newline at end of file
+</Tabs>
+
+## Approach 5: Shell Sort
+<Tabs>
+<TabItem value="py" label="Python">
+<SolutionAuthor name="@wingkwong"/>
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:362]: use your name

File: solutions/0000-0099/0088-merge-sorted-array-easy.md [hunk 2]
```diff
@@ -382,3 +382,183 @@ public:
...
+   
+4. After the right pointer reaches the end, decrease the value of the gap by setting it to ceil(current gap / 2).
+
+5. Repeat the loop until the gap becomes 0.
+
+6. Finally, after performing all the operations, you will have the merged sorted array.
+
+
+<Tabs>
+<TabItem value="cpp" label="C++">
```
Comment (wingkwong) [inline solutions/0000-0099/0088-merge-sorted-array-easy.md:406]: missing SolutionAuthor

File: solutions/0000-0099/0088-merge-sorted-array-easy.md [hunk 3]
```diff
@@ -382,3 +382,183 @@ public:
...
+
+```cpp
+class Solution {
```
Comment (wingkwong) [inline solutions/0000-0099/0088-merge-sorted-array-easy.md:409]: please format your code

File: tutorials/basic-topics/sorting/ShellSort.md [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+5. Perform a final pass of Insertion Sort with a gap of 1. This is a regular Insertion Sort that ensures the entire array is sorted correctly.
+
+Now, let's visualize how Shell Sort works with a simple example:
+
+<!-- We can see that Shell Sort is performed in the following way: -->
+
+Suppose we have an unsorted array `[6, 5, 3, 1, 8, 7, 2, 4]`. We'll perform Shell Sort on this array step by step:
+
+![image](../../../static/img/ShellSort.png)
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/ShellSort.md:51]: 1. please check out CONTRIBUTION guide to see how to upload image.
2. please cut the images to put the corresponding part after each step

**Review by wingkwong — CHANGES_REQUESTED** [review]
Gap sequence is a important section. also knuth sequence and sedgewick seqeunce should be included as the running time depends on it heavily.

File: tutorials/basic-topics/sorting/ShellSort.md (hunk 4 above)
Comment (Saishreekouda) [reply tutorials/basic-topics/sorting/ShellSort.md:51]: Okay, I will make these changes

**Comment by Saishreekouda** [comment]
@wingkwong made the changes

**Comment by Saishreekouda** [comment]
@wingkwong any more changes to be made?

File: tutorials/basic-topics/sorting/ShellSort.md [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+
+4. Repeat steps 2 and 3 until the gap becomes 1.
+
+### Step 5: Final Pass
+
+5. Perform a final pass of Insertion Sort with a gap of 1. This is a regular Insertion Sort that ensures the entire array is sorted correctly.
+
+Now, let's visualize how Shell Sort works with a simple example:
+
+<!-- We can see that Shell Sort is performed in the following way: -->
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/ShellSort.md:71]: double check if this line should be commented or not.

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. change the filename. `ShellFort.md` -> `shell-sort.md`
2. remove static/img. 
3. for formulas / variables / numbers, please use latex. e.g. use `$` instead.

**Comment by Saishreekouda** [comment]
@wingkwong done!

**Comment by Saishreekouda** [comment]
@wingkwong are there any changes to be made?

**Review by wingkwong — CHANGES_REQUESTED** [review]
I couldn't start the app. Please run it locally to verify first.

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 6]
```diff
@@ -11,4 +10,463 @@ keywords:
   - shell sort
 ---
 
-<TutorialAuthors names="@TBC"/>
+<TutorialAuthors names="@<Saishreekouda>"/>
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:13]: remove `<>`, i.e. `@Saishreekouda`

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 7]
```diff
@@ -1,8 +1,7 @@
 ---
 title: 'Shell Sort'
-description: ''
-hide_table_of_contents: true
-draft: true
+description: 'Sorting algorithm using a gap concept'
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:3]: need a better description

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 8]
```diff
@@ -11,4 +10,463 @@ keywords:
...
+        j++;
+    }
+    }
+};
+```
+</TabItem>
+<TabItem value="py" label="Python3">
+
+```py
+class Solution(object):
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:325]: this is not python3. Please use python3 one instead - starting with

```py
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        
```

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 9]
```diff
@@ -11,4 +10,463 @@ keywords:
...
+
+For each value of the gap, the inner loop can run for a maximum of $n+m$ times.
+
+**Space Complexity** : $O(1)$
+
+**Efficiency Consideration**
+
+In this question we can use brute approach of merging 2 sorted arrays using extra space but for optimal solution where we do not consider any extra space Shell Sort gives us an efficient solution.
+
+export const suggestedProblems = [
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:444]: for each `problemName`, please add a space between `-`.  Also pad 0, e.g. `912` -> `0912`, `75` -> `0075`, etc.

**Review by wingkwong — CHANGES_REQUESTED** [review]

**Comment by Saishreekouda** [comment]
> I couldn't start the app. Please run it locally to verify first.

@wingkwong It's due to the images that I added on the github editor but that is giving error when i locally start it. I'm trying to fix this, I'd be glad if you could help.

**Comment by wingkwong** [comment]
> > I couldn't start the app. Please run it locally to verify first.
> 
> @wingkwong It's due to the images that I added on the github editor but that is giving error when i locally start it. I'm trying to fix this, I'd be glad if you could help.

Why would you use `` tag? You can directly use the result generated by Github Editor, e.g.

```
![image](https://github.com/wingkwong/leetcode-the-hard-way/assets/35857179/b88821ec-96d8-436a-b91d-bf9508bc87bf)
```

**Comment by Saishreekouda** [comment]
@wingkwong fixed it.

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 10]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+            for i in range(gap, n):
+                temp = nums[i]
+                j = i
+                while j >= gap and nums[j - gap] > temp:
+                    nums[j] = nums[j - gap]
+                    j -= gap
+                nums[j] = temp
+            gap //= 2
+
+    def sortArray(self, nums):
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:185]: not a python3 format.

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 11]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+
+**Efficiency Consideration**
+
+For very large arrays, Shell Sort's space complexity of $O(n)$ can be inefficient, as it allocates extra memory. In such cases, when memory usage is a critical concern, you may consider alternative sorting algorithms like Quick Sort, which typically requires less additional space.
+
+Shell Sort remains a reliable sorting algorithm for moderately sized arrays, and its performance can be optimized by choosing an appropriate gap sequence.
+
+
+
+## Example 2: [88 - Merge Sorted Array](https://leetcode.com/problems/merge-sorted-array)
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:245]: `88 - Merge Sorted Array` -> `0088 - Merge Sorted Array`.

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 12]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+
+> Given two integer arrays, sorted in non-decreasing order, and two integers representing the number of elements in both arrays.
+> Merge both arrays into a single array sorted in non-decreasing order.
+
+### Algorithm
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:250]: no need to have such header. please remove it.

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 13]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+
+1. Calculate the gap value for merging the two arrays. The gap is determined as $\lceil \frac{{\text{size of arr1[]} + \text{size of arr2[]}}}{{2}} \rceil$.
+
+2. Initialize two pointers: the left pointer at index 0 and the right pointer at index (left + gap).
+
+3. Perform the following steps for each gap until the gap becomes 0:
+   - Inside a loop that continues until the right pointer reaches the end (i.e., index $n + m$), handle three different cases:
+     - If the left pointer is inside arr1[] and the right pointer is in arr2[], compare arr1[left] and arr2[right - n]. If arr1[left] > arr2[right - n], swap them.
+     - If both pointers are in arr2[], compare arr1[left - n] and arr2[right - n]. If arr1[left - n] > arr2[right - n], swap them.
+     - If both pointers are in arr1[], compare arr1[left] and arr2[right]. If arr1[left] > arr2[right], swap them.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:260]: apply latex.

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 14]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+**Time Complexity**: $O((n+m) \cdot \log(n+m))$
+
+The time complexity of the algorithm for merging two sorted arrays of sizes $n$ and $m$ is determined as follows:
+- The gap ranges from $n+m$ to 1, and at each step, the gap is divided by 2.
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:433]: latex on `1` and `2`

File: tutorials/basic-topics/sorting/shell-sort.md (hunk 9 above)
Comment (wingkwong) [reply tutorials/basic-topics/sorting/shell-sort.md:444]: space one not resolved. e.g. `0912-Sort an Array` -> `0912 - Sort an Array`

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 15]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+**Efficiency Consideration**
+
+In this question we can use brute approach of merging 2 sorted arrays using extra space but for optimal solution where we do not consider any extra space Shell Sort gives us an efficient solution.
+
+export const suggestedProblems = [
+    {
+        "problemName": "0912-Sort an Array",
+        "difficulty": "Medium",
+        "leetCodeLink": "https://leetcode.com/problems/sort-an-array/",
+        "solutionLink": "../../../solutions/0900-0999/0912-sort-an-array-medium"
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:449]: incorrect link. `0921-` is not required

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 16]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+        "problemName" : "0075-Sort Colors",
+        "difficulty" : "Medium",
+        "leetCodeLink" : "https://leetcode.com/problems/sort-colors/",
+        "solutionLink" : "../../../solutions/0000-0099/sort-colors-medium"
+    }, 
+    {
+        "problemName" : "0088-Merge Sorted arrays",
+        "difficulty" : "Medium",
+        "leetCodeLink" : "https://leetcode.com/problems/merge-sorted-array/",
+        "solutionLink" : "../../../solutions/0000-0099/0088-merge-sorted-array-easy"
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:467]: incorrect link. `0088-` is not required

File: tutorials/basic-topics/sorting/shell-sort.md [hunk 17]
```diff
@@ -11,4 +10,468 @@ keywords:
...
+    }
+]
+
+<Table title="Suggested Problems" data={suggestedProblems} />
+
+
+
+
+
+
```
Comment (wingkwong) [inline tutorials/basic-topics/sorting/shell-sort.md:477]: remove unnecessary line breaks ..

File: solutions/0000-0099/0088-merge-sorted-array-easy.md [hunk 18]
```diff
@@ -382,3 +382,183 @@ public:
...
+
+3. Perform the following steps for each gap until the gap becomes 0:
+   - Inside a loop that continues until the right pointer reaches the end (i.e., index n + m), handle three different cases:
+     - If the left pointer is inside arr1[] and the right pointer is in arr2[], compare arr1[left] and arr2[right - n]. If arr1[left] > arr2[right - n], swap them.
+     - If both pointers are in arr2[], compare arr1[left - n] and arr2[right - n]. If arr1[left - n] > arr2[right - n], swap them.
+     - If both pointers are in arr1[], compare arr1[left] and arr2[right]. If arr1[left] > arr2[right], swap them.
+   
+4. After the right pointer reaches the end, decrease the value of the gap by setting it to ceil(current gap / 2).
+
+5. Repeat the loop until the gap becomes 0.
```
Comment (wingkwong) [inline solutions/0000-0099/0088-merge-sorted-array-easy.md:400]: apply latex

File: solutions/0000-0099/0088-merge-sorted-array-easy.md [hunk 19]
```diff
@@ -382,3 +382,183 @@ public:
...
+            j++;
+        }
+    }
+};
+```
+</TabItem>
+<TabItem value="py" label="Python3">
+
+```py
+class Solution(object):
```
Comment (wingkwong) [inline solutions/0000-0099/0088-merge-sorted-array-easy.md:462]: not python3 format.

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 20]
```diff
@@ -354,4 +354,60 @@ class Solution:
         return ans
 ```
 </TabItem>
-</Tabs>
\ No newline at end of file
+</Tabs>
+
+## Approach 5: Shell Sort
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:359]: 1. broken page
2. add explanation of this approach.

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. remove all commented ``.
2. please verify all pages once before re-requesting review.

File: tutorials/basic-topics/sorting/shell-sort.md (hunk 9 above)
Comment (wingkwong) [reply tutorials/basic-topics/sorting/shell-sort.md:436]: What is the reason to delete 0912?

Comment (Saishreekouda) [reply tutorials/basic-topics/sorting/shell-sort.md:436]: As it is already given in the examples

**Comment by Saishreekouda** [comment]
@wingkwong please check

**Comment by wingkwong** [comment]
@Saishreekouda will review tomorrow. next time please click re-request review. Sometimes I may miss it after getting flooded by notifications.

**Pull request merged**